Clock Correction in msmod

dricki opened the pull request:
We added clock correction as an msmod option. Data providers could use this to correct instrument times with a known drift, as is the case for ocean-bottom seismometers where the instrument-reference offset is measured before and after deployment.

The code in msmod.c is commented and generally follow the description in https://github.com/FDSN/OBS-standards/blob/main/other/msmod_drift_addition.md with the following caveats:

1) Computation of time is done in microseconds and we log the time with higher precision than in the design document
2) Logging is done via ms_log() function and since it is not initialized in msmod, the log output goes to the default stdout
3) Because of 2) Error 5) (Log Exists) is not relevant
4) Error 3 is currently implemented without a suggestion how to correct the the configuration file
5) Error 4 is not implemented because the original version of msmod strips time correction before we enter processmod() function 
	  /* Revert time to uncorrected value if correction was applied during unpacking */
	  if ( msr->fsdh->time_correct != 0 && ! (msr->fsdh->act_flags & 0x02) )
	    {
	      msr->starttime = msr_starttime_uc (msr);
	    }
So if the user inputs time-corrected input file and tries to do the same time correction, the input and output files will be identical and no error and error messages are printed out. If this behavior is not what we want  I suggest Chad modifies msr_starttime_uc() function.

dricki committed:
First modifications of the code for clock correction

dricki committed:
Adding clock configuration structure

dricki committed:
Completed clock configuration structure input

dricki committed:
Completed coding and testing of Clock Correction feature

chad-earthscope commented:
Thanks @dricki 
Also tagging @WayneCrawford

Three issues before I can consider merging.

1) No documentation in the man page for this new feature.  The extended help (-H option) has a bit of detail regarding the option but it doesn't look sufficient for user to understand the features of the clock correction field.  Extended help addition is (which also looks to have minor formatting issues):
```Console
  # The clock correction (--cc option) file format is: #
 type: {keyword} {parameters}
 # Instrument Time     Reference Time
 {instrument_time_0}   {reference_time_0}
 {instrument_time_1}   {reference_time_1}
 ....
```

2) Compiler warnings with the new code.  General rule is no compiler warnings with latest, or reasonably current, clang and gcc compilers.  The warnings:
```Console
cc -g -I../libmseed -c msmod.c
msmod.c:298:17: warning: using the result of an assignment as a condition without parentheses [-Wparentheses]
  298 |           if ( retcode = processmods (msr) )
      |                ~~~~~~~~^~~~~~~~~~~~~~~~~~~
msmod.c:298:17: note: place parentheses around the assignment to silence this warning
  298 |           if ( retcode = processmods (msr) )
      |                        ^
      |                (                          )
msmod.c:298:17: note: use '==' to turn this assignment into an equality comparison
  298 |           if ( retcode = processmods (msr) )
      |                        ^
      |                        ==
msmod.c:1984:11: warning: using floating point absolute value function 'fabs' when argument is of integer type [-Wabsolute-value]
 1984 |       if (fabs(cc_config->ref_time[i] - correction - cc_config->inst_time[i]) > SMALL)
      |           ^
msmod.c:1984:11: note: use function 'llabs' instead
 1984 |       if (fabs(cc_config->ref_time[i] - correction - cc_config->inst_time[i]) > SMALL)
      |           ^~~~
      |           llabs
```

3) The new code is a large addition relative to the main code.  It makes code review difficult and will make future maintenance harder.  New code should be in separate file.

dricki commented:
Hi, Chad -Ok, all three comments are reasonable and I will be happy to implement. Wayne, do you feel like writing text for man page or let me do it?Best regards -IlyaSent from my iPhoneOn Sep 4, 2025, at 11:40, Chad Trabant ***@***.***> wrote:﻿chad-earthscope left a comment (EarthScope/msmod#2)
Thanks @dricki
Also tagging @WayneCrawford
Three issues before I can consider merging.

No documentation in the man page for this new feature.  The extended help (-H option) has a bit of detail regarding the option but it doesn't look sufficient for user to understand the features of the clock correction field.  Extended help addition is (which also looks to have minor formatting issues):

  # The clock correction (--cc option) file format is: #
 type: {keyword} {parameters}
 # Instrument Time     Reference Time
 {instrument_time_0}   {reference_time_0}
 {instrument_time_1}   {reference_time_1}
 ....

Compiler warnings with the new code.  General rule is no compiler warnings with latest, or reasonably current, clang and gcc compilers.  The warnings:

cc -g -I../libmseed -c msmod.c
msmod.c:298:17: warning: using the result of an assignment as a condition without parentheses [-Wparentheses]
  298 |           if ( retcode = processmods (msr) )
      |                ~~~~~~~~^~~~~~~~~~~~~~~~~~~
msmod.c:298:17: note: place parentheses around the assignment to silence this warning
  298 |           if ( retcode = processmods (msr) )
      |                        ^
      |                (                          )
msmod.c:298:17: note: use '==' to turn this assignment into an equality comparison
  298 |           if ( retcode = processmods (msr) )
      |                        ^
      |                        ==
msmod.c:1984:11: warning: using floating point absolute value function 'fabs' when argument is of integer type [-Wabsolute-value]
 1984 |       if (fabs(cc_config->ref_time[i] - correction - cc_config->inst_time[i]) > SMALL)
      |           ^
msmod.c:1984:11: note: use function 'llabs' instead
 1984 |       if (fabs(cc_config->ref_time[i] - correction - cc_config->inst_time[i]) > SMALL)
      |           ^~~~
      |           llabs

The new code is a large addition relative to the main code.  It makes code review difficult and will make future maintenance harder.  New code should be in separate file.


—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>









To follow ISTI news and updates please subscribe to our newsletter "the isti letter" at  https://www.isti.com/newsletter-sign-up.

WayneCrawford commented:
@dricki , I can make a stab at the help text, but would prefer that you add it in once I'm done, as I'm not very familiar/comfortable with pull requests.

dricki commented:
Dear Wayne As long as you write the text and simply email me I will do the restBest regards -IlyaSent from my iPhoneOn Sep 4, 2025, at 17:00, Wayne Crawford ***@***.***> wrote:﻿WayneCrawford left a comment (EarthScope/msmod#2)
@dricki , I can make a stab at the help text, but would prefer that you add it in once I'm done, as I'm not very familiar/comfortable with pull requests.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>









To follow ISTI news and updates please subscribe to our newsletter "the isti letter" at  https://www.isti.com/newsletter-sign-up.

WayneCrawford commented:
Here is my recommendation.  I'm putting it here so that Chad can verify it's not too long before you go to the bother of inserting it:

```c
      fprintf (stderr,
           "\n"
	       "  # Clock correction (--cc option) #\n"
	       " Sets timecorrection and modifies starttime in every record according\n"
	       " to a clock drift specified in the clock correction input file.\n"
	       "\n"
	       "The clock correction input file format is:"
	       " type: {keyword} [{parameters}]\n"
	       " # Instrument Time     Reference Time\n"
	       " {instrument_time_0}   {reference_time_0}\n"
	       " {instrument_time_1}   {reference_time_1}\n"
	       " ....\n"
	       "\n"
	       "The times in each column must monotonically increase and the\n"
	       "{instrument_time}s must cover the time range of the miniSEED file(s).\n"
	       "Time format is yyyy-mm-ddTHH:MM:SS(.FFFFF)Z.\n"
	       "Comment lines start with '#' and have no effect on processing.\n"
	       "Possible 'type' lines are:\n"
	       "type: piecewise_linear\n:
	       "    shifts instrument_time to reference_time for each provided value,\n"
	       "    linearly interpolates in between\n"
	       "type: cubic_spline\n"
	       "    shifts instrument_time to reference_time for each provided value,\n"
	       "    cubic spline interpolation in between\n"
	       "type: polynomial a0 a1 a2 a3...\n"
	       "    sets corrected_time = instrument_time_0 + a0 + a1*delta + a2*delta**2 + ...,\n"
	       "    where delta = instrument_time - instrument_time_0.\n"
           "    {instrument_time_n} and {reference_time_n} are used to validate results\n"
	       "\n"
	       ""Also outputs a text file named {CCFILENAME}.log, with the columns:\n"
	       "    RecNo: the record number\n"
	       "    Instrument time: original instrument time (ISO8601)\n"
	       "    Corrected to reference: corrected time (ISO8601)\n"
	       "    Corrected-Instrument: corrected time minus instrument time (s)\n"
	       "    Instrument-sync_inst[0]: instrument time - instrument_time_0 (s)\n"
	       "\n");
```

chad-earthscope commented:
[Edit with more clarity]

The command line help message `-h` should include the option in the listing (as it is now) and in the extended message with `-H` a brief description, similar to what is there now but formatted better.  

The [doc/msmod.1](https://github.com/EarthScope/msmod/blob/main/doc/msmod.1) page should include:
* addition of the `--cc` option to the argument list
* the long description of the correction file format in it's own section.

dricki committed:
Fixed warnings from gcc -Wall

WayneCrawford commented:
With that in mind, I would recommend shortening the `-H` description to:

```c
      fprintf (stderr,
           "\n"
	       " # Clock correction (--cc option) #\n"
	       "     Sets timecorrection and modifies starttime in every record according\n"
	       "     to the specified clock drift.\n"
	       "\n"
	       "    Clock correction input file format:"
	       "\n"
	       "    {type_line}\n"
	       "    {instrument_time_0}   {reference_time_0}\n"
	       "    {instrument_time_1}   {reference_time_1}\n"
	       "    ....\n"
	       "\n"
	       "    Possible {type_line}s:\n"
	       "    type: piecewise_linear\n:
	       "    type: cubic_spline\n"
	       "    type: polynomial a0 a1 a2 a3...\n");
	       "\n"
	       "    Time format: yyyy-mm-ddTHH:MM:SS(.FFFFF)Z.\n"
```

dricki committed:
Updated man pages and -H help option

dricki committed:
Moved clock correction code into clockcorr.c/clockcorr.h file

dricki commented:
@chad-earthscope and @WayneCrawford -
I think I have completed the requests from Chad:
1) Compiler warnings
2) Man page and help update
3) Moving the new code to clockcorr.c file

.md file is not updated, versions and dates are not updated
Also makefiles.wat and makefile.win files need update to include the new source file

Please review the pull request

WayneCrawford commented:
Hi Ilya

I will download and test the code.  Could you copy the changes you put in msmod.1 to msmod.md?  I tried but it requires me to fork.

Cheers
Wayne

dricki commented:
Hi, Wayne -

Chad told me that he has a script to convert man2md and he will do it

Cheers -

Ilya

On 9/22/2025 1:48 PM, Wayne Crawford wrote:
> *WayneCrawford* left a comment (EarthScope/msmod#2) 
> <https://github.com/EarthScope/msmod/pull/2#issuecomment-3318196643>
>
> Hi Ilya
>
> I will download and test the code. Could you copy the changes you put 
> in msmod.1 to msmod.md? I tried but it requires me to fork.
>
> Cheers
> Wayne
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/EarthScope/msmod/pull/2#issuecomment-3318196643>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ABNDCECG5KBAV6CZ47AHQH33T7HZJAVCNFSM6AAAAACCMINMJSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTGMJYGE4TMNRUGM>.
> You are receiving this because you were mentioned.Message ID: 
> ***@***.***>
>
>
> /To follow ISTI news and updates please subscribe to our newsletter 
> "the isti letter" at https://www.isti.com/newsletter-sign-up 
> <https://www.isti.com/newsletter-sign-up>./
>
-- 








*To follow ISTI news and updates please subscribe to our newsletter 
"the isti letter" at  https://www.isti.com/newsletter-sign-up 
<https://www.isti.com/newsletter-sign-up>.*

WayneCrawford commented:
OK great,

I've started looking at the code.  First, on my system, I get the following compile warning:

```console
cc -g -I../libmseed -c clockcorr.c
clockcorr.c:697:11: warning: absolute value function 'abs' given an argument of type 'hptime_t' (aka 'long long') but has parameter of type 'int' which may cause truncation of value [-Wabsolute-value]
  697 |       if (abs(cc_config->ref_time[i] - correction - cc_config->inst_time[i]) > SMALL)
      |           ^
clockcorr.c:697:11: note: use function 'llabs' instead
  697 |       if (abs(cc_config->ref_time[i] - correction - cc_config->inst_time[i]) > SMALL)
      |           ^~~
      |           llabs
1 warning generated.
```

Then, when I run ``msmod -H`` I get the following "CC" section:
```console
# Clock correction (--cc option) #
     Sets timecorrection and modifies starttime in every record according
     to the specified clock drift.

    Clock correction input file format:
    {type_line}
    {instrument_time_0}   {reference_time_0}
    {instrument_time_1}   {reference_time_1}
    ....

    Possible {type_line}s:
    type: piecewise_linear
:    type: cubic_spline
    type: polynomial a0 a1 a2 a3...

    Time format: yyyy-mm-ddTHH:MM:SS(.FFFFF)Z.
```

which I would change to (new stuff in "**", if Chad approves):

```console
# Clock correction (--cc option) #
     Sets **record header** timecorrection and **starttime fields** according
     to the **original starttime and** the specified clock drift.

    Clock correction input file format:
    **type:** {type_value}
    {instrument_time_0}   {reference_time_0}
    {instrument_time_1}   {reference_time_1}
    ....

    Possible {type_value}s are:
        piecewise_linear
        cubic_spline
        polynomial a0 a1 a2 a3...

    **{*_time_*}** format: yyyy-mm-ddTHH:MM:SS(.FFFFF)Z
```

Finally, when I run

```
> msmod --cc clock_correct_cubic.txt test_30sph.mseed
```

I get

```console
ERROR Unknown option: --cc
```

I get the same error for 
```
> msmod --cc clock_correct_cubic.txt test_30sph.mseed -o test_30sph_cubic.mseed
```

Either I'm not using the command line correctly, or there is some bug that's not accepting "--cc".  I will look to see if I find the reason but you will probably be a lot faster than me!

Cheers
Wayne

dricki committed:
Fixes based on comments from Wayne: 1 warning and -H option

WayneCrawford commented:
Complilation warning is gone.
``--cc`` option is recognized (maybe I made an error during testing?)

cubic_spline file read gives error because it expects ``cubic-spline`` instead of ``cubic_spline``.  ``cubic_spline`` is how it's written in the docs and is consistent with the ``piecewise_linear`` type, please change to ``cubic_spline``.  Also, error message is not very informative:
```console
ERROR: Badly formatted input file: line 1
ERROR reading CC configuration: 'clock_correct_cubic.txt'
```
maybe could say "unknown type on type line: xxx"

Otherwise, all the different test files (linear1, linear2, polynomial and cubic (with "cubic-spline")) give output, I will now check if the output matches the specs.

WayneCrawford commented:
I checked the code with 9 different input files:
- linear correction with two sync times
- linear correction with three sync times
- cubic spline correction with three sync times
- polynomial correction with three sync times
- 2 "bad" files in which the first sync is after the data start: one by 1 second, one by 1 month
- 2 "bad" files in which the last sync is before the data end: one by 1 second, one by 1 month
- a "bad" file in which the sync times are not increasing with each increasing row.
- a "bad" file in which the reference synchronization time does not match the given polynomial

## Major bug: miniSEED record header start times are not modified

In each new record header, the time correction value is entered and the "time correction applied" flag is set, but the start time is the same as before.  Here's an example, pruned from ``msi -p`` using ``grep -e "start time" -e "time correction"``:
```console
    mseed input file, last headers...
             start time: 2022,331,01:12:00.000000
        time correction: 0
             start time: 2022,340,05:14:00.000000
        time correction: 0
             start time: 2022,349,09:16:00.000000
        time correction: 0
             start time: 2022,358,13:18:00.000000
        time correction: 0
    mseed output file, last headers...
             start time: 2022,331,01:12:00.000000
        time correction: -12713
             start time: 2022,340,05:14:00.000000
        time correction: -13313
             start time: 2022,349,09:16:00.000000
        time correction: -13913
             start time: 2022,358,13:18:00.000000
        time correction: -14512
```

## Drift correction values

- The linear correction worked fine (within a microsecond of my results)
- The cubic spline correction worked fine (but the "type" field should be "cubic_spline", not "cubic-spline"
- The polynomial works ok, although it seems to confuse the "instrument" and "reference" times:
    - the ``sync_instrument[0]`` in the log file seems to actually be the ``sync_reference[0]``
    - the ``CORRECTED-REFERENCE`` in the polynomial time check error message is actually CORRECTED-INSTRUMENT

## Output log file

- There is a strange large space in "Corrected to      reference" on the first line
- Seconds are shown to microsecond precision (miniSEED2 only has 100-microsecond precision, but this
   is probably fine as miniSEED3 has higher precision)

## Output miniSEED file
- Time corrections are truncations of the "Corrected - Instrument" rather than rounding to the closest value (for example "-14384" for "-1.438470"
- Time corrections are not applied to the Record start time

## Error catching and signalling

### Time checking on polynomial
Works, but the ``CORRECTED-REFERENCE`` value is actually the CORRECTED minus the INSTRUMENT

### Catching a too-late first sync time
Works for the 1 month offset, not for the 1 second offset.
The 1-second offset subsequently fails because of interpolation bounds, but the message should be the same as for the 1-month offset

### Catching a too-early last sync time
Works for the 1-month offset, not for the 1-second offset. 
No subsequent fail, probably because the last second starts more than a second before the end of data, but it should be caught anyway.

### Catching an out-of-order sync time
Works.

## Attached files

The zipped miniSEED file I ran the tests on:
[test_30sph.mseed.zip](https://github.com/user-attachments/files/22518753/test_30sph.mseed.zip)

The "bad" test files I used:
[clock_correct_bad_end_very.txt](https://github.com/user-attachments/files/22518699/clock_correct_bad_end_very.txt)
[clock_correct_bad_end.txt](https://github.com/user-attachments/files/22518700/clock_correct_bad_end.txt)
[clock_correct_bad_order.txt](https://github.com/user-attachments/files/22518701/clock_correct_bad_order.txt)
[clock_correct_bad_polynomial.txt](https://github.com/user-attachments/files/22518702/clock_correct_bad_polynomial.txt)
[clock_correct_bad_start_very.txt](https://github.com/user-attachments/files/22518703/clock_correct_bad_start_very.txt)
[clock_correct_bad_start.txt](https://github.com/user-attachments/files/22518704/clock_correct_bad_start.txt)

The code I used to run the tests:
[run_tests.sh](https://github.com/user-attachments/files/22518931/run_tests.sh)

The outputs of the different tests:
[test_results.md](https://github.com/user-attachments/files/22518694/test_results.md)

I can also provide the "standard" test clock_correct files and their expected outputs (I had to slightly modify them from what I provided initially, in order to match the msmod output)

dricki committed:
Replaced cubic-spline config token with cubic_spline

dricki committed:
Improved error-related logging

dricki committed:
After Wayne's comment fixed a bug where corrected time was not writtin to the output file

dricki committed:
Fixed a bug mixing inst amd ref time in log of clock correction

dricki committed:
Fixed a bug in error log for polynomial case

chad-earthscope committed:
Reformatting -H output

chad-earthscope committed:
Sync markdown manual with man page

chad-earthscope commented:
Please ping me when this passes muster @WayneCrawford and I'll merge into the main repo.

The plan is to merge this into a `1.x` branch and then I'll port the code to a new major version that includes miniSEED v2 and v3 support.

WayneCrawford commented:
I recompiled the code (no errors or warnings) and ran the code on my test suite. For "good" input files, the outputs are all good and most of what's left is formatting, except for 1) the creation of miniSEED and log files after errors and 2) the identification of bad time bounds in ``piecewise_linear`` and ``cubic_spline`` input files:

## Overview of test results

- For good input files, the outputs are now all good (piecewise_linear, cubic_spline, and polynomial)
- For the "error" input files, the error messages are generally good, but:
     - they still generate  (sometimes empty) miniseed and log files generated, which could lead to confusion. 
       Any errors should not generate output files (unless it's a log file with only the error messages)
     - Some of the error messages are not preceded by "ERROR": I think they should be, for clarity.
- the log file still has a strange long space in the middle of the third header ("``Corrected to    reference``") and that header starts before the column it represents
- In the Markdown documentation, there seem to be two extraneous lines in the "Clock correction input file format" description:
  <img width="1056" height="287" alt="Screenshot 2025-10-02 at 12 10 19" src="https://github.com/user-attachments/assets/7bdcf1fd-f506-406b-98d0-2c870d5a3b53" />
  
I'm also still not sure what criteria you use to decide whether the date bounds cover the data time limits.  Allowing a little slack seems to work ok for a slightly out of range end time, because the last record starts before the end of data (but by how much is variable).  But you will always get an interpolation error for a bad start time, so you might as well nip it in the bud for consistency.  I understand that this could create a problem for polynomials, which have a completely different relationship to the "instrument" and "reference" times. I suggest that you implement strict checking for piecewise_linear and cubic_spline, but not for polynomial

## Individual test evaluations

### clock_correct_bad_end_very.txt

Gives an error, but I think the first line should be preceded by ERROR, too:

```
Testing clock_correct_bad_end_very.txt...
Data ends after last instrument time (by 537238.50 seconds).
ERROR, processing Clock Correction failed
ERROR modifying:
```

Also still creates a mseed file and a log file, should not

### clock_correct_bad_end.txt

No error, what is the tolerance (1 record?)

### clock_correct_bad_order.txt

Good ERROR message, no mseed file created, **empty log file created**

### clock_correct_bad_polynomial.txt

Good ERROR message, no mseed file created, **empty log file created**

### clock_correct_bad_start_very.txt

Good ERROR message, **empty mseed file and 1-line log file created**

log file:
```
XX_STA__LXX, 000001, D, 4096, 6601 samples, 0.008333333333 Hz, 2022,001,00:00:00.000000
```

### clock_correct_bad_start.txt

Error is created, but first line does not start with ERROR:

```
Time in MSEED header is out of interpolation bounds.
ERROR, processing Clock Correction failed
ERROR modifying:
  Error processing test_30sph.mseed: Generic error
```

**empty mseed file and 1-line log file created**

log file:
```
XX_STA__LXX, 000001, D, 4096, 6601 samples, 0.008333333333 Hz, 2022,001,00:00:00.000000
```

### clock_correct_cubic.txt

My original input file still had ``cubic-spline``, gave correct error, no mseed file created, but **empty log file created**

With corrected input file, Log and mseed files look good

### clock_correct_linear1.txt

Log and mseed files look good

### clock_correct_linear2.txt

Log and mseed files look good
(last significant digit is still not correctly rounded, for example -1.451288 seconds becomes -14512 time correction, but do we care?):

```
     40  2022-12-24T13:18:00.000000   2022-12-24T13:17:58.548712      -1.451288           30892680.000000
    mseed input file, last headers...
             start time: 2022,358,13:18:00.000000
        time correction: 0
    mseed output file, last headers...
             start time: 2022,358,13:17:58.548700
        time correction: -14512
```

### clock_correct_polynomial.txt

Log and mseed files look good

WayneCrawford commented:
Hi Ilya

Any updates/questions?

Cheers
Wayne

dricki commented:
Dear @WayneCrawford  -
I emailed you on November 5, 2025, but it looks like it went nowhere ( Wayne Crawford <notifications@github.com).
Indeed, I do have comments and questions and I repeat here. Also, I am very sorry for the delay

> I recompiled the code (no errors or warnings) and ran the code on my test suite. For "good" input files, the outputs are all good and most of what's left is formatting, except for 1) the creation of miniSEED and log files after errors and 2) the identification of bad time bounds in piecewise_linear and cubic_spline input files:

Now, comments and questions:

1a) Creation of miniSEED files: I left a mechanism of creation (and not removing in case of an error) without changes.

This is how original Chad's msmod operates:

a) Opens up an output file

b) Processes each MSEED record individually, checking for errors in each MSEED record

c) Exits at the first erroneous MSEED or when there is no more records left

The output file is not being deleted in case of an error.

I can add the code to remove the output file in case of an error. I ill do it only for clock correction option.

It also can be expanded for all options in the program if Chad wants. For now please confirm that I should make sure

output file is removed for Clock Correction option in case of an error

1b) Creation of log files. 

As I mention error checking is done simultaneously with processing input MSEED records. Thus, logging is done per record

and before error is found (end-out-of-bounds) processing of many records can be done. So logging simply shows what was done: we processed records and at some point ran into the problem. I suggest we do not leave empty log files. In the beginning of processing we can create log message: Clock Correction processing started and in the end we can log Clock Correction processing failed or Clock  Correction processing succeed. So each log file will have at least two messages.

2) Time bounds are set to 2 seconds in all correction types 

3) Overview of test results

> For the "error" input files, the error messages are generally good, but:
> 
>     they still generate (sometimes empty) miniseed and log files generated, which could lead to confusion.
>     Any errors should not generate output files (unless it's a log file with only the error messages)

 See my comments above 

> Some of the error messages are not preceded by "ERROR": I think they should be, for clarity.
I fixed it  
> the log file still has a strange long space in the middle of the third header ("Corrected to reference") and that header starts before the column it represents
Could you please send me an example of this strange long space. I maybe run the program differently 

> In the Markdown documentation, there seem to be two extraneous lines in the "Clock correction input file format" description:
[Screenshot.2025-10-02.at.12.10.19.png (view on web)](https://github.com/user-attachments/assets/7bdcf1fd-f506-406b-98d0-2c870d5a3b53)

I think I now fixed it 

> I'm also still not sure what criteria you use to decide whether the date bounds cover the data time limits. Allowing a little slack seems to work ok for a slightly out of range end time, because the last record starts before the end of data (but by how much is variable). But you will always get an interpolation error for a bad start time, so you might as well nip it in the bud for consistency. I understand that this could create a problem for polynomials, which have a completely different relationship to the "instrument" and "reference" times. I suggest that you implement strict checking for piecewise_linear and cubic_spline, but not for polynomial

2 seconds 

> 
> Good ERROR message, empty mseed file and 1-line log file created
> 
> log file:
> 
> XX_STA__LXX, 000001, D, 4096, 6601 samples, 0.008333333333 Hz, 2022,001,00:00:00.000000

This  logging done by the Chad's code. It demonstrates that we work with the particular MSEED record

> clock_correct_linear2.txt
> 
> Log and mseed files look good
> (last significant digit is still not correctly rounded, for example -1.451288 seconds becomes -14512 time correction, but do we care?):
  
I  think this rounding error is not relevant 

==============================

To summarize:
1) I suggest I add the code which removes the output file in case of an error which is effectively similar to not creating it but does not require too much changes of the msmod code. Need a confirmation from you
2) I suggest we allow creation of log files even in case of an error but make sure empty log files are not created. Need a confirmation from you
3) I suggest we allow 2 second 2 bounds for all algorithms. Need a confirmation from you
4) I am waiting for  an example of the strange long space in the middle of the third header ("Corrected to reference") from you.
5) Other issues are resolved in the development version

After the first 4 issues are resolved I update git repository for the round of testings (hopefully, final)

WayneCrawford commented:
Dear @dricki

Thank you for your rapid reponse.  I didn't see the first one, will try to make
up for lost time!
I'll start by answering your comments, questions (except number 2) and your summary.  I'll answer number 2 last:
I think it's the only remaining "sticking point".

## Questions and comments:
### 1a

I will defer to @chad-earthscope on this: the most important thing is that
   we do not disobey msmod's core philosophy.
   
### 1b

Your solution sounds good to me.

### 3

I agree that your answer to 1b addresses this.  I'll ask @chad-earthscope to weigh in on this.

## Summary

1. This sounds like a good solution, but I will check with @chad-earthscope to make sure that this is compatible with his
   vision for msmod
2. I confirm this solution
3. I think this is more complex than an n-second delay: see below.
4. I have put my entire test directory [here](https://filesender.renater.fr/?s=download&token=890b82ee-52b8-4cac-b43f-e3aec65007f8): a relevant output file is `results_20251002/clock_correct_linear1.txt.test_out.log`, for example.
5. Excellent.

## Comment 2

2 seconds sounds fine to me, but I don't think that the current version
acheives this.  I tested 4 "bad time range" configurations:
   
1. `clock_correct_bad_start.txt`: The first sync provided is 1 **second** after the data start, the result is:
       - Run error is
       
         ```
         Time in MSEED header is out of interpolation bounds.
         ERROR, processing Clock Correction failed
         ERROR modifying:
              Error processing test_30sph.mseed: Generic error
         ```
       - log file is one line: ``XX_STA__LXX, 000001, D, 4096, 6601 samples, 0.008333333333 Hz, 2022,001,00:00:00.000000`` (Chad's code)
       - Empty miniSEED file created
        
2. `clock_correct_bad_start_very.txt`:The first sync provided is 1 **month** after the data start
       - Run error is
       
         ```
          Data starts before first instrument time (by 2678401.00 seconds).
          ERROR, processing Clock Correction failed
          ERROR modifying:
              Error processing test_30sph.mseed: Generic error
         ```
       - log file is one line: ``XX_STA__LXX, 000001, D, 4096, 6601 samples, 0.008333333333 Hz, 2022,001,00:00:00.000000`` (Chad's code)
       - Empty miniSEED file created
3. `clock_correct_bad_end.txt`:The last sync provided is 1 **day** before the data end
       - No run error
       - Full log file (header + 40 records)`
       - Full miniSEED file created
4. `clock_correct_bad_end_very.txt`: The last sync provided is 1 **month** before the data end
       - Run error is
       
         ```
         Data ends after last instrument time (by 537238.50 seconds).
         ERROR, processing Clock Correction failed
         ERROR modifying:
             Error processing test_30sph.mseed: Generic error
         ```
       - log file is header + 37 records plus line: ``XX_STA__LXX, 000038, D, 4096, 6601 samples, 0.008333333333 Hz, 2022,340,05:14:00.000000``
       - Incomplete miniSEED file created
   
So a first sync 1 **second** after the first data still creates an error (but with a different error message
than if it were 3 seconds after), but a last sync 1 **day** before the last data does not create any error.
I'm guessing that your code checks for a 2 second offset with respect to the record starts, but the interpolation
code fails if the current record is not within the sync time bounds.  This means that, if the data start time is the slightest bit after the first sync, the code fails, but gives a different message depending on whether it was your "offset checker"  (2 seconds) or the interpolation (no tolerance) that triggered the error.  The same thing probably holds for the start of the last record (which is not the end of the data, and in my example the records are a few
days long).

Ideally, the code should give the same error message in any case where: 1) the first sync is after the first data, or
2) the last sync is before the last data.  It seems to me that the simplest
way to do this would be to check from the start whether the data start or data end are outside of the sync time bounds.  This would also avoid Chad's error message.

Note that this test should only be applied to the `piecewise-linear` and `cubic-spline` methods, as the `polynomial` sync times are for verification only.

dricki committed:
Fixes after test run by Wayne

dricki commented:
Dear @WayneCrawford -
I made several changes to the code in the clock_correction branch: https://github.com/dricki/msmod/tree/clock_correction
Could you please run a your test script and let me know if all right and what not. thank you. Hopefully, this will bring us very close to the release

WayneCrawford commented:
Hi Ilya,

Thank you, I will run those tests.

Regards
Wayne
  
 _______________________________Wayne CrawfordCNRS Researcher
Bureau 341
Institut de Physique du Globe de Paris
1, rue Jussieu
75238 Paris Cedex 05
+33 6 14 42 77 70


  
On 1 December 2025 at 04:04:06, Ilya Dricker ***@***.***) wrote:
 dricki left a comment (EarthScope/msmod#2)

Dear @WayneCrawford -

I made several changes to the code in the clock_correction branch: https://github.com/dricki/msmod/tree/clock_correction

Could you please run a your test script and let me know if all right and what not. thank you. Hopefully, this will bring us very close to the release


—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you were mentioned.Message ID: ***@***.***>

WayneCrawford commented:
Hi Ilya,

I ran my tests on your latest code, and it looks really close!  My "good input" tests all conform and the "bad input" tests are much better (no miniSEED file created, uniform nomenclature for errors.  However, if it's not too difficult, I'd like these "bad input" log files to be a little more descriptive.  Here's an example for each case (bad_start, bad_end, bad_order and bad_polynomial)

## bad\_start

The output is:

```
Clock Correction processing started
ERROR, Clock Correction processing failed
XX_STA__LXX, 000001, D, 4096, 6601 samples, 0.008333333333 Hz, 2022,001,00:00:00.000000
```

I'd prefer that the second to last line were:

```
Clock Correction processing started
ERROR, Clock Correction processing failed: record start is before correction file's instrument start time
XX_STA__LXX, 000001, D, 4096, 6601 samples, 0.008333333333 Hz, 2022,001,00:00:00.000000
```

## bad\_end

The output is:

```
Clock Correction processing started
# RecNo  Instrument time            Corrected to    reference     Corrected-Instrument    Instrument-sync_inst[0]
      1  2022-01-01T00:00:00.000000   2022-01-01T00:00:00.000000       0.000000                  0.000000
      2  2022-01-10T04:02:00.000000   2022-01-10T04:01:59.962220      -0.037780             792120.000000
...
     39  2022-12-15T09:16:00.000000   2022-12-15T09:15:58.564344      -1.435656           30100560.000000
ERROR, Clock Correction processing failed
XX_STA__LXX, 000040, D, 4096, 5362 samples, 0.008333333333 Hz, 2022,358,13:18:00.000000
```

I'd prefer that the second to last line were:

```
ERROR, Clock Correction processing failed: record start is after correction file's instrument end time
```

## bad\_order

The output is:

```
Clock Correction processing started
```

without any error!  I'd prefer that it were

```
Clock Correction processing started
ERROR, Clock Correction processing failed: correction file times are not monotonically increasing
```

## bad\_polynomial

The output is:

```
Clock Correction processing started
```

without any error!  I'd prefer that it were

```
Clock Correction processing started
ERROR, Clock Correction processing failed: correction file times do not match provided polynomial
```

If it's easier to put the additional information on another line, that would work too.  

Regards
Wayne

dricki committed:
Improved logging in clockcorr.c

WayneCrawford commented:
I just ran my tests on the latest version of msmod.  The log files and miniSEED output files all look good.

**There is one bug**: the fails on bad start and end times return code 0.  I think you can very quickly fix this.

# Overview of log files for error messages

I like the logs, they are all concise but clear:

## `clock_correct_bad_polynomial`

```
Clock Correction processing started
ERROR, polynomial does not generate reference corrected times
INSTRUMENT_TIME             |   REFERENCE_TIME            |    CORRECTED_TIME           | CORRECTED-REFERENCE (s)
--------------------------- | --------------------------- | --------------------------- | -----------------------
2023-01-01T00:00:01.400000  | 2023-01-01T00:00:00.000000  | 2022-12-31T23:59:59.900082  |  -0.099918
```

## `clock_correct_bad_order`

```
Clock Correction processing started
ERROR, non-increasing instrument times: time line {#2}
```

## `clock_correct_bad_start`

```
Clock Correction processing started
ERROR, time in MSEED header is out of interpolation bounds.
ERROR, Clock Correction processing failed
XX_STA__LXX, 000001, D, 4096, 6601 samples, 0.008333333333 Hz, 2022,001,00:00:00.000000
```

## `clock_correct_bad_start_very`

```
Clock Correction processing started
ERROR, data starts before first instrument time (by 2678401.00 seconds).
ERROR, Clock Correction processing failed
XX_STA__LXX, 000001, D, 4096, 6601 samples, 0.008333333333 Hz, 2022,001,00:00:00.000000
```

## `clock_correct_bad_end`

```
Clock Correction processing started
# RecNo  Instrument time            Corrected to    reference     Corrected-Instrument    Instrument-sync_inst[0]
      1  2022-01-01T00:00:00.000000   2022-01-01T00:00:00.000000       0.000000                  0.000000
      2  2022-01-10T04:02:00.000000   2022-01-10T04:01:59.962220      -0.037780             792120.000000
[I cut out 3-37 here]
     38  2022-12-06T05:14:00.000000   2022-12-06T05:13:58.602124      -1.397876           29308440.000000
     39  2022-12-15T09:16:00.000000   2022-12-15T09:15:58.564344      -1.435656           30100560.000000
ERROR, data ends after last instrument time (by 86398.50 seconds).
ERROR, Clock Correction processing failed
XX_STA__LXX, 000040, D, 4096, 5362 samples, 0.008333333333 Hz, 2022,358,13:18:00.000000
```

## `clock_correct_bad_end_very`

```
Clock Correction processing started
# RecNo  Instrument time            Corrected to    reference     Corrected-Instrument    Instrument-sync_inst[0]
      1  2022-01-01T00:00:00.000000   2022-01-01T00:00:00.000000       0.000000                  0.000000
      2  2022-01-10T04:02:00.000000   2022-01-10T04:01:59.958703      -0.041297             792120.000000
[I cut out 3-34 here]
     35  2022-11-08T17:08:00.000000   2022-11-08T17:07:58.595884      -1.404116           26932080.000000
     36  2022-11-17T21:10:00.000000   2022-11-17T21:09:58.554586      -1.445414           27724200.000000
ERROR, data ends after last instrument time (by 537118.50 seconds).
ERROR, Clock Correction processing failed
XX_STA__LXX, 000037, D, 4096, 6601 samples, 0.008333333333 Hz, 2022,331,01:12:00.000000
```


# Keeping incomplete miniSEED files

Note that if, in the future, we decide to output the miniSEED file in case of failure in a record after the first one, only the "*_bad_end*" correction files would keep a miniSEED file, as the others fail on or before the first record.

# Overview of test results

Below is a stripped down version of the test results containing, for each run:

    - the command line run
    - any output from the command line
    - the return code

```console
>  ./msmod --cc clock_correct_bad_end_very.txt test_30sph.mseed -o clock_correct_bad_end_very.txt.mseed > clock_correct_bad_end_very.txt.test_out.log
ERROR modifying:
  Error processing test_30sph.mseed: Generic error
return code = 0
```

```console
>  ./msmod --cc clock_correct_bad_end.txt test_30sph.mseed -o clock_correct_bad_end.txt.mseed > clock_correct_bad_end.txt.test_out.log
ERROR modifying:
  Error processing test_30sph.mseed: Generic error
return code = 0
```

```console
>  ./msmod --cc clock_correct_bad_order.txt test_30sph.mseed -o clock_correct_bad_order.txt.mseed > clock_correct_bad_order.txt.test_out.log
ERROR reading CC configuration: 'clock_correct_bad_order.txt'
return code = 1
```

```console
>  ./msmod --cc clock_correct_bad_polynomial.txt test_30sph.mseed -o clock_correct_bad_polynomial.txt.mseed > clock_correct_bad_polynomial.txt.test_out.log
ERROR reading CC configuration: 'clock_correct_bad_polynomial.txt'
return code = 1
```

```console
>  ./msmod --cc clock_correct_bad_start_very.txt test_30sph.mseed -o clock_correct_bad_start_very.txt.mseed > clock_correct_bad_start_very.txt.test_out.log
ERROR modifying:
  Error processing test_30sph.mseed: Generic error
return code = 0
```

```console
>  ./msmod --cc clock_correct_bad_start.txt test_30sph.mseed -o clock_correct_bad_start.txt.mseed > clock_correct_bad_start.txt.test_out.log
ERROR modifying:
  Error processing test_30sph.mseed: Generic error
return code = 0
```

```console
>  ./msmod --cc clock_correct_cubic.txt test_30sph.mseed -o clock_correct_cubic.txt.mseed > clock_correct_cubic.txt.test_out.log
return code = 0
```

```console
>  ./msmod --cc clock_correct_linear1.txt test_30sph.mseed -o clock_correct_linear1.txt.mseed > clock_correct_linear1.txt.test_out.log
return code = 0
```

```console
>  ./msmod --cc clock_correct_linear2.txt test_30sph.mseed -o clock_correct_linear2.txt.mseed > clock_correct_linear2.txt.test_out.log
return code = 0
```

```console
>  ./msmod --cc clock_correct_polynomial.txt test_30sph.mseed -o clock_correct_polynomial.txt.mseed > clock_correct_polynomial.txt.test_out.log
return code = 0
```

dricki committed:
Fixed return code of the program for errors: improved error logging

WayneCrawford commented:
The latest build passes all of my tests.  Details below:

====================================================
There are 10 "correction" files:

- 4 good:
    - `clock\_correct\_cubic`
    - `clock\_correct_linear1`
    - `clock\_correct\_linear2`
    - `clock\_correct\_polynomial`
- 6 bad:
    - 2 with bad values in the file
        - `clock\_correct\_bad\_polynomial`: check values do not match polynomial
        - `clock\_correct\_bad\_order`: times are out of order
    - 4 with time values inconsistent with the data range
        - `clock\_correct\_bad_start`
        - `clock\_correct\_bad\_start\_very`
        - `clock\_correct\_bad\_end`
        - `clock\_correct\_bad\_end\_very`

Each good file also has a model output log file.

## Were "good" files properly processed?
      
- output mseed file
    - [x] miniSEED files created for all 4 good inputs?
    - [x] miniSEED files match predicted?
        - [x] Same number of samples as input?
            - ```
              > msi -T ../test_30sph.mseed
              
                 Source                Start sample             End sample        Hz  Samples
              XX_STA__LXX       2022,001,00:00:00.000000 2023,001,00:00:00.000000 0.00833 262801
              ```
            -
              ```console
              > msi -T clock_correct_cubic.txt.mseed
                 Source                Start sample             End sample        Hz  Samples
              XX_STA__LXX       2022,001,00:00:00.000000 2022,365,23:59:58.561500 0.00833 262801
              
              > msi -T clock_correct_linear1.txt.mseed
                 Source                Start sample             End sample        Hz  Samples
              XX_STA__LXX       2022,001,00:00:00.000000 2022,365,23:59:58.530600 0.00833 262801

              > msi -T clock_correct_linear2.txt.mseed
                 Source                Start sample             End sample        Hz  Samples
              XX_STA__LXX       2022,001,00:00:00.000000 2022,365,23:59:58.548700 0.00833 262801

              > msi -T clock_correct_polynomial.txt.mseed
                 Source                Start sample             End sample        Hz  Samples
              XX_STA__LXX       2021,365,23:59:59.999000 2022,365,23:59:58.558500 0.00833 262801
              ```

        - [x] Same size as input?
            - ```
              > ls -l ../*.mseed
              -rw-r--r--  1 crawford  admin  163840 Oct  9  2024 ../test_30sph.mseed
              
              > ls -l *.mseed
              -rw-r--r--  1 crawford  admin  163840 Jan 22 09:30 clock_correct_cubic.txt.mseed
              -rw-r--r--  1 crawford  admin  163840 Jan 22 09:30 clock_correct_linear1.txt.mseed
              -rw-r--r--  1 crawford  admin  163840 Jan 22 09:30 clock_correct_linear2.txt.mseed
              -rw-r--r--  1 crawford  admin  163840 Jan 22 09:30 clock_correct_polynomial.txt.mseed
	          ```

        - [x] Offsets at end consistent with expected (msi)?
        - [x] Time correction field consistent with record start times?
- output log file
    - [x] log files created for all of good files?
    - [x] log files match predicted?
        - only acceptable differences is least significant digit (1 microsecond, different rounding?)
- agreement between log and miniSEED files
    - [x] Plot of time corrections matches?
- [x] return code 0?
  
## Were "bad" files properly processed?
  
- [x] no mseed file?
- output log files as expected?
    - [x] `log` files created?
    - [x] `log` file contents consistent and clear?
        - ``bad_end`` and ``bad_end_very`` output several records, before ending with:

          ```
          ERROR, data ends after last instrument time (by 86398.50 seconds).
          ERROR, Clock Correction processing failed
          XX_STA__LXX, 0000xx, D, 4096, xxxx samples, 0.008333333333 Hz, 2022,jjj,HH:MM:SS
          ```
        - ``bad_order`` log file:
        
          ```
          Clock Correction processing started
          ERROR, non-increasing instrument times: time line {#2}
          ```
        - ``bad_polynomial`` log file:
        
          ```
          Clock Correction processing started
          ERROR, polynomial does not generate reference corrected times
          INSTRUMENT_TIME             |   REFERENCE_TIME            |    CORRECTED_TIME           | CORRECTED-REFERENCE (s)
          --------------------------- | --------------------------- | --------------------------- | -----------------------
          2023-01-01T00:00:01.400000  | 2023-01-01T00:00:00.000000  | 2022-12-31T23:59:59.900082  |  -0.099918
          ```
        - ``bad_start`` log file:
        
          ```
          Clock Correction processing started
          ERROR, time in MSEED header is out of interpolation bounds.
          ERROR, Clock Correction processing failed
          XX_STA__LXX, 000001, D, 4096, 6601 samples, 0.008333333333 Hz, 2022,001,00:00:00.000000
          ```
        - ``bad_start_very`` log file:
        
          ```
          Clock Correction processing started
          ERROR, data starts before first instrument time (by 2678401.00 seconds).
          ERROR, Clock Correction processing failed
          XX_STA__LXX, 000001, D, 4096, 6601 samples, 0.008333333333 Hz, 2022,001,00:00:00.000000
          ```


- [x]  return code != 0?

## Are standard error messages consistent and clear?

- [x] No message for "good" inputs?
- [x] Clear message for "bad" inputs?
    - ```
      ERROR modifying:
        Error processing test_30sph.mseed: Clock Correction error
      ```
      for ``bad_start``, ``bad_start_very``, ``bad_end`` and  ``bad_end_very``

    - ```
      ERROR reading CC configuration: '{filename}'
      ```
      for ``bad_polynomial`` and ``bad_order``

chad-earthscope committed:
Update markdown manual from man page

chad-earthscope committed:
Update version to 1.3.0 and remove trailing whitespace
